tm2py/emme/manager.py

```diff
@@ -0,0 +1,238 @@
+            dimensions: dictionary of the specified dimensions to set.
+        """
+        dims = emmebank.dimensions
+        new_dims = dims.copy()
+        new_dims.update(dimensions)
+        if dims != new_dims:
+            change_dimensions = self.tool(
+                "inro.emme.data.database.change_database_dimensions"
+            )
+            change_dimensions(new_dims, emmebank, keep_backup=False)
+
+    def modeller(self, emme_project: EmmeDesktopApp = None) -> EmmeModeller:
+        """Initialize and return Modeller object.
+
+        If Modeller has not already been initialized it will do so on
+        specified Emme project, or the first Emme project opened if not provided.
+        If already initialized Modeller will reference whichever project was used
+        first.
+
+        Args:
+            emme_project: open 'Emme Desktop' application (inro.emme.desktop.app)
+        """
+        # pylint: disable=E0611, E0401, E1101
+        try:
+            return _m.Modeller()
+        except AssertionError as error:
+            if emme_project is None:
+                if self._project_cache:
+                    emme_project = next(iter(self._project_cache.values()))
+                else:
+                    raise Exception(
+                        "modeller not yet initialized and no cached Emme project,"
+                        " emme_project arg must be provided"
+                    ) from error
+            return _m.Modeller(emme_project)
+
+    def tool(self, namespace: str):
+        """Return the Modeller tool at namespace."""
+        return self.modeller().tool(namespace)
+
+    @staticmethod
+    @_context
+    def temp_attributes_and_restore(
+        scenario: EmmeScenario, attributes: List[List[str]]
+    ):
+        """Create temp extra attribute and network field, and backup values and state and restore.
+
+        Allows the use of temporary attributes which may conflict with existing attributes.
+        The temp created attributes are deleted at the end, and if there were pre-existing
+        attributes with the same names the values are restored.
+
+        Note that name conflicts may still arise in the shorthand inheritance systems
+        for the network hierarchy tree (@node attribute reserves -> @nodei, @nodej, etc,
+        see Emme help Network calculations for full list) which will raise an error in the
+        Emme API.
+
+        Args:
+            scenario: Emme scenario object
+            attributes: list of attribute details, where details is a list of 3 items
+                for extra attributes and 4 for network fields: domain, name, description[, atype]
+        """
+        attrs_to_delete = []
+        fields_to_delete = []
+        attrs_to_restore = dict(
+            (d, []) for d in ["NODE", "LINK", "TURN", "TRANSIT_LINE", "TRANSIT_SEGMENT"]
+        )
+        for details in attributes:
+            domain, name, desc = details[:3]
```

Comment:
is the length/ordering of each item in attributes guaranteed? (just want to make sure this line wont bomb if not)

Reply:
It is as doc'd, used a few places so I can't see a big issue but does require reading the docstring or looking at an example carefully.